### PR TITLE
Fix custom gate

### PR DESCRIPTION
**Pull request by hJaffaliColibritd**

### CODE REVIEW

File: tests/core/instruction/gates/test_custom_gate.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    # building the expected state vector
+    exp_state_vector = m[:, 0]
+    for _ in range(0, t):
+        exp_state_vector = np.kron(np.array([1, 0]), exp_state_vector)
+    for _ in range(t + n_gate, n_circ):
+        exp_state_vector = np.kron(exp_state_vector, np.array([1, 0]))
+
+    execution_ibm_statevector = run(c, IBMDevice.AER_SIMULATOR).state_vector
+    execution_aws_statevector = run(c, AWSDevice.BRAKET_LOCAL_SIMULATOR).state_vector
+    execution_qlm_statevector = run(c, ATOSDevice.MYQLM_PYLINALG).state_vector
```
Comment (JulienCalistoTD) [inline tests/core/instruction/gates/test_custom_gate.py:48]: We might be able to do a batch result

File: tests/core/instruction/gates/test_gate.py [hunk 2]
```diff
@@ -71,6 +71,7 @@ def test_power(gate: Gate, pow: float, result_matrix: npt.NDArray[np.complex64])
     assert np.allclose(gate.power(pow).to_matrix(), result_matrix)
 
 
+@pytest.mark.skip
```
Comment (JulienCalistoTD) [inline tests/core/instruction/gates/test_gate.py:74]: why skip this test?

File: mpqp/core/instruction/gates/gate_definition.py [hunk 3]
```diff
@@ -123,15 +123,27 @@ def __init__(self, definition: Matrix, disable_symbol_warn: bool = False):
                 "Matrices defining gates have to be unitary. It is not the case"
                 f" for\n{definition}"
             )
+        elif not is_power_of_two(definition.shape[0]):
```
[the remark this answers is not in the document]
Comment (hJaffaliColibritd) [reply mpqp/core/instruction/gates/gate_definition.py:126]: Is unitary only check if the operator is unitary, not if it is suitable for qubits (which is only the case if the square matrix has dimension that is power of two), so I would say we need this extra check

File: tests/core/instruction/gates/test_gate.py (hunk 2 above)
Comment (hJaffaliColibritd) [reply tests/core/instruction/gates/test_gate.py:74]: Because the feature of tensor product between gates is not fully implemented, so the test is ready but will not succeed, so we skip it

File: mpqp/core/instruction/gates/gate_definition.py [hunk 4]
```diff
@@ -123,15 +123,27 @@ def __init__(self, definition: Matrix, disable_symbol_warn: bool = False):
                 "Matrices defining gates have to be unitary. It is not the case"
```
[the remark this answers is not in the document]
Comment (hJaffaliColibritd) [reply mpqp/core/instruction/gates/gate_definition.py:123]: You we replace the symbolic by specific values, and if the matrix is not unitary it means that the symbolic one was also not ?

[the remark this answers is not in the document]
Comment (JulienCalistoTD) [reply mpqp/core/instruction/gates/gate_definition.py:123]: No, sometimes we have an expression without a variable, so can do a evalf. But if we have a variable, i think it's too difficult to know if it's a unitary.

**Review by JulienCalistoTD — APPROVED** [review]

File: mpqp/core/instruction/gates/gate_definition.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hJaffaliColibritd) [reply mpqp/core/instruction/gates/gate_definition.py:123]: Maybe a way to check is to compute the doc product between each line/column and then ask sympy to simplify it, and check that it is always zero. We also have to check that each line/colmun forms a vector of norm 1, which I think can also be done with symbolic calculus. Imma add a task on Trello.

File: mpqp/core/circuit.py [hunk 5]
```diff
@@ -133,6 +132,11 @@ def __init__(
         self._nb_qubits: int
         """Number of qubits of the circuit."""
 
+        self.gphase: Optional[float] = None
```
Comment (Henri-ColibrITD) [inline mpqp/core/circuit.py:135]: wouldn't it be simpler if it was 1 by default ? and not optional ?

File: mpqp/core/circuit.py [hunk 6]
```diff
@@ -913,14 +918,14 @@ def to_other_language(
                 cargs = []
 
                 if isinstance(instruction, CustomGate):
+                    # We reverse the target qubits to pass the unitary in the usual qubit ordering
+                    reversed_target = instruction.targets
+                    reversed_target.reverse()
                     new_circ.unitary(
                         instruction.to_other_language(),
-                        instruction.targets,
+                        reversed_target,
```
Comment (Henri-ColibrITD) [inline mpqp/core/circuit.py:926]:
```py
reversed(instruction.targets)
```

File: mpqp/core/circuit.py [hunk 7]
```diff
@@ -1045,16 +1052,78 @@ def to_qasm2(self) -> str:
...
         qiskit_circ = self.subs({}, remove_symbolic=True).to_other_language(
             Language.QISKIT
         )
         if TYPE_CHECKING:
             assert isinstance(qiskit_circ, QuantumCircuit)
 
-        from qiskit import qasm2
+        qiskit_circ.reverse_bits()
+
+        if all([not isinstance(inst, CustomGate) for inst in self.instructions]):
```
Comment (Henri-ColibrITD) [inline mpqp/core/circuit.py:1108]: why is this needed ? as I see it the following loop will cover this case, and it doesn't seem to save a lot of time

File: mpqp/core/circuit.py (hunk 6 above)
Comment (hJaffaliColibritd) [reply mpqp/core/circuit.py:926]: Indeed, we can do that. However, `reversed(...)` returns an iterator, is `list(reversed(instructions.targets))` a suitable solution?

File: mpqp/core/circuit.py (hunk 5 above)
Comment (hJaffaliColibritd) [reply mpqp/core/circuit.py:135]: Yes we can indeed, then we check if it is different from 1, we multiply the state_vector by it at the result time, if 1 we do nothing.

Comment (hJaffaliColibritd) [reply mpqp/core/circuit.py:135]: Done

File: mpqp/core/circuit.py (hunk 6 above)
Comment (hJaffaliColibritd) [reply mpqp/core/circuit.py:926]: Done

File: tests/core/instruction/gates/test_gate.py (hunk 2 above)
Comment (Henri-ColibrITD) [reply tests/core/instruction/gates/test_gate.py:74]: I implemented the tensor product i think, so we can run this test now

**Review by Henri-ColibrITD — CHANGES_REQUESTED** [review]
see questions

File: mpqp/core/circuit.py (hunk 7 above)
Comment (Henri-ColibrITD) [reply mpqp/core/circuit.py:1074]: the key point is that the two loops are not nested but one after the other, so it's `n+n` and not `n*n`, so I'd say it's not a big deal, but if you prefer to keep the shortcut I'm fine with it

Comment (hJaffaliColibritd) [reply mpqp/core/circuit.py:1074]: I introduced a boolean variable to check if wether we encountered a CustomGate in the circuit or not, and then dump the original or new qiskit circuit.

File: mpqp/execution/result.py [hunk 8]
```diff
@@ -258,6 +258,9 @@ def __init__(
                 )
             else:
                 self._state_vector = data
+                if job.circuit.gphase is not None:
```
[the remark this answers is not in the document]
Comment (hJaffaliColibritd) [reply mpqp/execution/result.py]: Done

File: mpqp/core/instruction/gates/custom_gate.py [hunk 9]
```diff
@@ -12,19 +17,49 @@
...
+
+    Note:
+        For the moment, only ordered and contiguous target qubits are allowed when instantiating a CustomGate.
+
     """
 
     def __init__(
         self, definition: UnitaryMatrix, targets: list[int], label: Optional[str] = None
     ):
         self.matrix = definition.matrix
```
[the remark this answers is not in the document]
Comment (hJaffaliColibritd) [reply mpqp/core/instruction/gates/custom_gate.py:50]: We can indeed. To limit the influence on the rest of the code, I propose to define an attribute `self.unitary` and put the definition, and then define a property `self.matrix` that returns the matrix of the unitary.

This may however make pyright scream a bit if he can't understand that the type of self.matrix is Matrix, so will require to add some type cheking.

I let you confirm the idea before I implement.

File: mpqp/core/instruction/gates/custom_gate.py [hunk 10]
```diff
@@ -12,19 +17,49 @@
...
 
     Args:
-        definition: The matrix (this is the only way supported to now) semantics of the gate.
+        definition: The UnitaryMatrix (only way supported for now) description of the gate.
         targets: The qubits on which the gate operates.
         label: The label of the gate. Defaults to None.
+
+    Raises:
+        ValueError: the target qubits must be contiguous and in order, and must match the size of the UnitaryMatrix
+
```
[the remark this answers is not in the document]
Comment (hJaffaliColibritd) [reply mpqp/core/instruction/gates/custom_gate.py:31]: Not mandatory, we just need to re-use the work you did to generate any matrix from a gate, playing with the targets etc.

However, I think this should be done on another branch, after merging this one, because the work may not be straightfoward, when dealing with qiskit qubit ordering and change in the size of the CustomGate.

I create a task in Trello.

File: mpqp/execution/result.py [hunk 11]
```diff
@@ -258,6 +258,9 @@ def __init__(
                 )
             else:
                 self._state_vector = data
+                if job.circuit.gphase != 0:
```
Comment (Henri-ColibrITD) [inline mpqp/execution/result.py:261]: is this check really needed ?

File: mpqp/core/circuit.py (hunk 7 above)
Comment (Henri-ColibrITD) [reply mpqp/core/circuit.py]: I removed this variable (with your permission) because it doesn't really has any role in the result

File: mpqp/core/instruction/gates/custom_gate.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Henri-ColibrITD) [reply mpqp/core/instruction/gates/custom_gate.py:50]: meh, not sure it's important enough to spend that energy. It's just that if one day we add other ways to define gates, it may be annoying to handle.

Maybe just a task on trello ?

**Review by Henri-ColibrITD — APPROVED** [review]
a last comment to resolve but overall I'm fine with the PR !

File: mpqp/execution/result.py (hunk 11 above)
Comment (hJaffaliColibritd) [reply mpqp/execution/result.py:261]: In case there is no custom gate (which I assume will be the most common usage), I think checking that the phase is null is less costful than calling `np.exp` independently of the phase.
